packages/flutter/lib/src/widgets/inherited_theme.dart

```diff
@@ -0,0 +1,146 @@
+///
+/// ```dart main
+/// void main() {
+///   runApp(MyApp());
+/// }
+/// ```
+///
+/// ```dart
+/// class MyAppBody extends StatelessWidget {
+///   @override
+///   Widget build(BuildContext context) {
+///     return GestureDetector(
+///       onTap: () {
+///         Navigator.of(context).push(
+///           MaterialPageRoute(
+///             builder: (BuildContext _) {
+///               // InheritedTheme.captureAll() saves references to themes that
+///               // are found above the context provided to this widget's build
+///               // method, notably the DefaultTextStyle defined in MyApp. The
+///               // context passed to the MaterialPageRoute's builder is not used,
+///               // because its ancestors are above MyApp's home.
+///               return InheritedTheme.captureAll(context, Container(
+///                 alignment: Alignment.center,
+///                 color: Theme.of(context).colorScheme.surface,
+///                 child: Text('Hello World'),
+///               ));
+///             },
+///           ),
+///         );
+///       },
+///       child: Center(child: Text('Tap Here')),
+///     );
+///   }
+/// }
+///
+/// class MyApp extends StatelessWidget {
+///   @override
+///   Widget build(BuildContext context) {
+///     return MaterialApp(
+///       home: Scaffold(
+///         // Override the DefaultTextStyle defined by the Scaffold.
+///         // Descendant widgets will inherit this big blue text style.
+///         body: DefaultTextStyle(
+///           style: TextStyle(fontSize: 48, color: Colors.blue),
+///           child: MyAppBody(),
+///         ),
+///       ),
+///     );
+///   }
+/// }
+/// ```
+/// {@end-tool}
+abstract class InheritedTheme extends InheritedWidget {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const InheritedTheme({
+    Key key,
+    @required Widget child,
+  }) : super(key: key, child: child);
+
+  /// Return a copy of this inherited theme with the specified [child].
+  ///
+  /// If the identical inherited theme is already visible from [context] then
+  /// just return the [child].
+  ///
+  /// This implementation for [TooltipTheme] is typical:
+  /// ```dart
+  /// Widget wrap(BuildContext context, Widget child) {
+  ///   final TooltipTheme ancestorTheme = context.ancestorWidgetOfExactType(TooltipTheme);
+  ///   return identical(this, ancestorTheme) ? child : TooltipTheme(data: data, child: child);
+  /// }
+  /// ```
+  Widget wrap(BuildContext context, Widget child);
+
+  /// Returns a widget that will [wrap] child in all of the inherited themes
+  /// which are visible from [context].
+  static Widget captureAll(BuildContext context, Widget child) {
+    assert(child != null);
+    assert(context != null);
+
+    final List<InheritedTheme> themes = <InheritedTheme>[];
+    context.visitAncestorElements((Element ancestor) {
```

Comment:
Maybe we can move this to framework.dart so it has access to `_inheritedWidgets`?

Reply:
I don't think that there's a real performance problem due to walking past all of the ancestors; however you're right that this could perform a little better if the code was in framework.dart. On the other hand, framework.dart is quite big; keeping InheritedTheme separate makes everything just a little easier to read.